pkg/innerring/internal/blockchain/blockchain.go

```diff
@@ -0,0 +1,383 @@
+	// LevelDB or BoltDB.
+	UseLevelDB bool
+}
+
+// Config configures Blockchain. All fields are required.
+type Config struct {
+	// Application level error listener. Blockchain writes any internal error to the
+	// channel. Error returns of functions (e.g. New) are not pushed. The channel
+	// should be regularly checked in order to prevent blocking. Any pop-up error
+	// does not allow the Blockchain to fully work, therefore, if it is detected,
+	// the blockchain should be stopped.
+	ErrorListener chan<- error
+
+	// NEO wallet of the node. The wallet is used by Consensus and Notary services.
+	// Corresponding private key be accessed via Blockchain.LocalKey.
+	Wallet config.Wallet
+
+	// Storage configuration.
+	Storage StorageConfig
+
+	// Identifier of the Neo network.
+	NetworkMagic netmode.Magic
+
+	// Initial committee staff.
+	Committee keys.PublicKeys
+}
+
+// Options tunes Blockchain's default behavior. All fields are optional: zero
+// (omitted) values don't change the behavior. Launched Blockchain should be
+// stopped.
+type Options struct {
+	// Writer of the Blockchain's logs and internal errors. By default, Blockchain
+	// doesn't write logs.
+	Logger *zap.Logger
+
+	// Network addresses to listen Neo P2P in 'host:port' format. By default, Neo
+	// P2P is not served.
+	P2pEndpoints []string
+
+	// List of nodes' addresses to communicate with over Neo P2P protocol in
+	// 'host:port' format. By default, node runs as standalone.
+	SeedNodes []string
+
+	// Network addresses to listen Neo RPC in 'host:port' format. By default, Neo
+	// RPC is not served.
+	RPCEndpoints []string
+
+	// Disable Notary service.
+	NotaryDisabled bool
+}
+
+// New returns new Blockchain configured by specified Config and Options.
+// Resulting Blockchain is ready to run.
+func New(cfg Config, opts Options) (*Blockchain, error) {
+	switch {
+	case cfg.Wallet.Path == "":
+		panic("missing wallet path")
+	case cfg.ErrorListener == nil:
+		panic("missing error channel")
+	case cfg.NetworkMagic == 0:
+		// actually zero magic is valid, but almost definitely forgotten
+		panic("missing network magic")
+	case len(cfg.Committee) == 0:
+		panic("empty committee")
+	}
+
+	if opts.Logger == nil {
+		opts.Logger = zap.NewNop()
+	}
+
+	nodeAcc, err := utilConfig.LoadAccount(cfg.Wallet.Path, "", cfg.Wallet.Password)
+	if err != nil {
+		return nil, fmt.Errorf("read node account: %w", err)
+	}
+
+	standByCommittee := make([]string, len(cfg.Committee))
+	for i := range cfg.Committee {
+		standByCommittee[i] = hex.EncodeToString(cfg.Committee[i].Bytes())
+	}
+
+	var cfgBase config.Config
+
+	cfgBaseProto := &cfgBase.ProtocolConfiguration
+	cfgBaseProto.Magic = cfg.NetworkMagic
+	cfgBaseProto.StandbyCommittee = standByCommittee
+	cfgBaseProto.ValidatorsCount = len(standByCommittee)
+	cfgBaseProto.TimePerBlock = time.Second
+	cfgBaseProto.SeedList = opts.SeedNodes
+	cfgBaseProto.VerifyTransactions = true
+	cfgBaseProto.P2PSigExtensions = !opts.NotaryDisabled
+
```

Comment:
`MaxTraceableBlocks` is also an interesting one that can differ from network to network.
`Hardforks` is important to have for future updates.